### ISSUE TITLE
Errors not being reported for kong plugin Opentelemetry

### ISSUE THREAD

**Issue by pawandhiman10** [Kong/kong#13776]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

3.7.1

### Current Behavior

We have installed this plugin for traces in Kong. 
Plugin [link](https://docs.konghq.com/hub/kong-inc/opentelemetry/)

We are sending traces to Datadog with this. But the traces are only for requests and errors are not captured at all.
<img width="1095" alt="Screenshot 2024-10-20 at 10 32 38 PM" src="https://github.com/user-attachments/assets/b4573cbb-8806-4c2f-a18a-23ab6b6ab43c">

### Expected Behavior

Errors should be captured and reported in the UI.

### Steps To Reproduce

Install this plugin using Helm.
We have attached this plugin to a service with the kong annotations: `konghq.com/plugins: opentelemetry`
Configuration reference for the kongplugin on the plugin definition page.
This is connected to otel configuration exposed by Datadog agent on http port 4318.

**Comment by ProBrian** [comment]
@pawandhiman10  can you attach you config file for the OTEL plugin? 
@samugi can you help to have a look at this issue?

**Comment by samugi** [comment]
in addition to the config file of the plugin, I would also like to clarify what is meant by "errors" here (i.e. are we talking about error logs during a certain plugin's execution? Or the error_log produced by Kong? Or something else?. @pawandhiman10 could you share **an example of the exact data** you would expect to be reported?

**Comment by samugi** [comment]
hello @pawandhiman10 
errors in traces are captured and reported by the OpenTelemetry plugin as span events using the `exception` event key and the `exception.message` attribute as defined in the [OpenTelemetry specification](https://opentelemetry.io/docs/specs/semconv/exceptions/exceptions-spans/#recording-an-exception).

I believe Datadog might be expecting slighgly different fields based on [this](https://docs.datadoghq.com/tracing/error_tracking/#use-span-tags-to-track-error-spans), are you using the [OTLP ingestion](https://docs.datadoghq.com/opentelemetry/interoperability/otlp_ingest_in_the_agent/?tab=host) with the datadog agent? In that case, I would expect the ingestion process of OTLP data to take care of parsing any errors and translating any fields as needed.

**Comment by pawandhiman10** [comment]
@samugi Errors mean 5xx status code errors here.
As per the screenshot earlier, it only displays the 2xx requests and not the 5xx. These are not actual errors but response is 5xx. This is missing.
Below screenshot gives an idea that 5xx is occuring regularly and also increasing sometimes, we are not able to see these numbers in datadog while the same is visible in the load balancer logs.
<img width="1381" alt="Screenshot 2024-10-21 at 12 36 34 PM" src="https://github.com/user-attachments/assets/065b07ab-e1a2-498a-89cc-29f532141a1c">

Yes, currently the ingestion is based on OLTP ingestion with Datadog agent. Could you point me to some reference docs here on how to get the errors (5xx) parsed?

**Comment by samugi** [comment]
@pawandhiman10 in that case, it might be expected. 500 response status codes are not always reported as errors. Today errors are reported for failures (exceptions) that occur during the execution of plugin phase handlers, errors returned by the internal http client and dns failures. 

Could you test with a 500 status code that originates from an exception in a plugin? This can be tested with a plugin that throws an explicit error like: `error("fail")` from its access phase. The expected behavior is for this to generate a span with name `kong.access.plugin.yourplugin` with error state and an event that contains the exception message. If Datadog is parsing things correctly, this should be displayed in the UI as well.

**Comment by samugi** [comment]
@pawandhiman10 what you describe is currently expected. Status codes are reported (whether they 2xx, 4xx, 5xx, etc), but spans are not being set the *error* state when this happens. Errors are currently reserved for actual errors occurring during the execution of plugins code (exceptions) and a few other scenarios that don't include specific response codes.

Would you be interested in setting an error state to your root span in case of 5xx status code from your upstream? If that is the case, such a change should probably be configurable, for example some might expect 4xx response codes from time to time, while others would consider them error conditions. 

We are always interested in improving/updating our tracing instrumentation, but I cannot guarantee this change in particular will be made.
I would recommend, to achieve this behavior, to write your own code (in a custom or serverless plugin) that applies this simple logic for your trace. You can use the [tracing](https://docs.konghq.com/gateway/latest/plugin-development/pdk/kong.tracing/) and the [response](https://docs.konghq.com/gateway/latest/plugin-development/pdk/kong.response/) PDK modules to achieve this.

**Comment by julianocosta89** [comment]
@pawandhiman10 if you have access to the collector configuration, you can use the [`transformprocessor`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor) to do that:

```
processors:
  transform:
    error_mode: ignore
    trace_statements:
      - context: span
        statements:
          - set (status.code, 2) where attributes["http.status_code"] >= 500
```

Here are the enums available, and the explanation why I've suggested `set (status.code, 2)`:
https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/contexts/ottlspan#enums

**Comment by pawandhiman10** [comment]
@julianocosta89 This is managed by Datadog agent currently without any collector config.

@samugi Could you also help me in setting `OTEL_TRACES_SAMPLER=always_on` with this plugin?
As of the error(5xx) reporting in Datadog, will experiment with the code. Thanks for that.

**Comment by samugi** [comment]
@pawandhiman10 at the moment the sampling strategy is not configurable for this plugin, what behavior in particular are you attempting to configure?

**Comment by pawandhiman10** [comment]
@samugi This is required I believe from the OTEL page to set the following env variables in Datadog. ref: [Link](https://opentelemetry.io/docs/languages/sdk-configuration/general/)

Enabling opentelemetry send traces in Datadog with probablistic sampling and we need to control the sampling rate of these traces. For that the following envs need to be set: [link](https://docs.datadoghq.com/opentelemetry/ingestion_sampling_with_opentelemetry/)
DD_APM_PROBABILISTIC_SAMPLER_ENABLED
DD_APM_PROBABILISTIC_SAMPLER_SAMPLING_PERCENTAGE
DD_APM_PROBABILISTIC_SAMPLER_HASH_SEED


I'm not quite sure on how this works but trying to reduce the ingested spans using this opentelemetry plugin.

**Comment by samugi** [comment]
@pawandhiman10 @julianocosta89 does the following describe your scenario?

> If you use a mixed setup of Datadog tracing libraries and OTel SDKs:
Probabilistic sampling will apply to spans originating from both Datadog and OTel tracing libraries.
If you send spans both to the Datadog Agent and OTel collector instances, set the same seed between Datadog Agent (DD_APM_PROBABILISTIC_SAMPLER_HASH_SEED) and OTel collector (hash_seed) to ensure consistent sampling.

Kong's OpenTelemetry plugin only supports head sampling today so your options for sampling, depending on your architecture choices, would be:

1. Rely solely on Kong's head sampling via [tracing_sampling_rate](https://docs.konghq.com/gateway/latest/reference/configuration/#tracing_sampling_rate)
2. Configure `tracing_sampling_rate` to 1 (sample all) and do tail sampling on the Datadog Agent
3. Configure `tracing_sampling_rate` to 1 (sample all), use an OpenTelemetry collector as a sink to receive Kong's tracing data, then configure probabilistic sampling in the OTel collector and the Datadog Agent, as suggested in the documentation you have shared

**Comment by julianocosta89** [comment]
So setting `tracing_sampling_rate` to 1, is the same as `OTEL_TRACES_SAMPLER=always_on`

**Comment by samugi** [comment]
that's right @julianocosta89 1 stands for 100%, i.e. all samples will be traced and reported.

**Comment by julianocosta89** [comment]
> So setting `tracing_sampling_rate` to 1, is the same as `OTEL_TRACES_SAMPLER=always_on`

@pawandhiman10 that's what you are looking for
